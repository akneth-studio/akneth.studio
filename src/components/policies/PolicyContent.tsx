'use client'

import ReactMarkdown from 'react-markdown'
import rehypeRaw from 'rehype-raw'
import { Badge } from 'react-bootstrap'
import Link from 'next/link'
import styles from '@/styles/Privacy.module.scss'

interface PolicyContentProps {
    h1Text: string;
    content: string;
    lastUpdated: string;
}

export default function PolicyContent({ h1Text, content, lastUpdated }: PolicyContentProps) {
    return (
        <article className={`${styles.article} p-3 p-md-5`}>
            <h1 className='page-title'>{h1Text}</h1>
            <div className='text-center' id='last-updated'>
                <Badge
                    bg='secondary'
                    text='light'
                    as='p'
                    style={{ fontSize: '0.9rem' }}
                    className='mx-auto px-3 py-2 fw-normal lh-base shadow-sm'
                >
                    Data ostatniej aktualizacji: <br className='d-block d-md-none py-1' />
                    <span className='fw-bold'>{lastUpdated}</span>
                </Badge>
            </div>
            <hr />
            <ReactMarkdown
                rehypePlugins={[rehypeRaw]}
                components={{
                    a: ({ href, children, title, ...props }) => {
                        const isInternal = href?.startsWith('/')
                        const ariaLabel = title || (typeof children === 'string' ? children : undefined)
                        if (isInternal) {
                            return (
                                <Link href={href ?? '#'} {...props} title={title} aria-label={ariaLabel}>
                                    {children}
                                </Link>
                            )
                        }
                        return (
                            <a href={href} target="_blank" rel="noopener noreferrer" {...props} title={title} aria-label={ariaLabel}>
                                {children}
                            </a>
                        )
                    }
                }}
            >
                {content}
            </ReactMarkdown>
        </article>
    );
}