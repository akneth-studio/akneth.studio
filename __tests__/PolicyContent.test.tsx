import React from 'react';
import { render, screen } from '@testing-library/react';
import PolicyContent from '../src/components/policies/PolicyContent';
import '@testing-library/jest-dom';

// Mock next/link for testing purposes
jest.mock('next/link', () => {
  return ({ children, href }: { children: React.ReactNode; href: string }) => {
    // eslint-disable-next-line jsx-a11y/anchor-has-content
    return <a href={href}>{children}</a>;
  };
});

const mockProps = {
  h1Text: 'Test Heading',
  content: `
This is some test content.
Here is an [internal link](/internal-page).
And here is an [external link](https://example.com).
  `,
  lastUpdated: '2024-01-01',
};

describe('PolicyContent component', () => {
  it('renders the policy content with all props and creates a snapshot', () => {
    const { container } = render(<PolicyContent {...mockProps} />);

    // Snapshot test
    expect(container).toMatchSnapshot();

    // Check for heading
    const heading = screen.getByRole('heading', {
      level: 1,
      name: /test heading/i,
    });
    expect(heading).toBeInTheDocument();

    // Check for last updated date
    // The text is split by a <br /> tag, so we check for the parts separately.
    expect(screen.getByText(/Data ostatniej aktualizacji:/i)).toBeInTheDocument();
    expect(screen.getByText('2024-01-01')).toBeInTheDocument();

    // Check for markdown content
    expect(screen.getByText(/This is some test content./i)).toBeInTheDocument();
  });

  it('renders internal links correctly', () => {
    render(<PolicyContent {...mockProps} />);
    const internalLink = screen.getByText('internal link');
    expect(internalLink).toBeInTheDocument();
    expect(internalLink).toHaveAttribute('href', '/internal-page');
    // It should not have target="_blank"
    expect(internalLink).not.toHaveAttribute('target', '_blank');
  });

  it('renders external links correctly', () => {
    render(<PolicyContent {...mockProps} />);
    const externalLink = screen.getByText('external link');
    expect(externalLink).toBeInTheDocument();
    expect(externalLink).toHaveAttribute('href', 'https://example.com');
    expect(externalLink).toHaveAttribute('target', '_blank');
    expect(externalLink).toHaveAttribute('rel', 'noopener noreferrer');
  });

  it('renders correctly when optional props are missing', () => {
    // Pass an empty string for lastUpdated to satisfy the required prop type
    // and test the component's ability to handle falsy values gracefully.
    const propsWithEmptyDate = {
      h1Text: 'Another Heading',
      content: 'Some other content without links.',
      lastUpdated: '',
    };
    render(<PolicyContent {...propsWithEmptyDate} />);

    expect(screen.getByRole('heading', { name: /another heading/i })).toBeInTheDocument();
    expect(screen.getByText(/some other content without links./i)).toBeInTheDocument();

    // Check that the last updated date is NOT rendered
    expect(screen.queryByText(/ostatnia aktualizacja/i)).not.toBeInTheDocument();
  });
});
