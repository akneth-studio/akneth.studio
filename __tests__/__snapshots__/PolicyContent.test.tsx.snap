// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`PolicyContent component renders the policy content with all props and creates a snapshot 1`] = `
<div>
  <article
    class="article p-3 p-md-5"
    data-testid="policy-content"
  >
    <h1
      class="page-title"
    >
      Test Heading
    </h1>
    <div
      class="text-center"
      id="last-updated"
    >
      <p
        class="mx-auto px-3 py-2 fw-normal lh-base shadow-sm badge text-light bg-secondary"
        style="font-size: 0.9rem;"
      >
        Data ostatniej aktualizacji: 
        <br
          class="d-block d-md-none py-1"
        />
        <span
          class="fw-bold"
        >
          2024-01-01
        </span>
      </p>
    </div>
    <hr />
    <p>
      This is some test content.
Here is an 
      <a
        aria-label="internal link"
        href="/internal-page"
        node="[object Object]"
      >
        internal link
      </a>
      .
And here is an 
      <a
        aria-label="external link"
        href="https://example.com"
        node="[object Object]"
        rel="noopener noreferrer"
        target="_blank"
      >
        external link
      </a>
      .
    </p>
  </article>
</div>
`;
